codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoMiddlewareGenerator.java

```diff
@@ -0,0 +1,201 @@
+                                .build())
+                        .build())
+                .handlerType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.SerializeHandler")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .build();
+    }
+
+    public static GoMiddlewareGenerator newDeserializeMiddleware(String identifier) {
+        return builder()
+                .identifier(identifier)
+                .handleMethodName("HandleDeserialize")
+                .inputType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.DeserializeInput")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .outputType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.DeserializeOutput")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .handlerType(SymbolReference.builder()
+                        .symbol(SymbolUtils.createValueSymbolBuilder("middleware.DeserializeHandler")
+                                .addReference(SymbolUtils.createNamespaceReference(GoDependency.SMITHY_MIDDLEWARE))
+                                .build())
+                        .build())
+                .build();
+    }
+
+    public void writeMiddleware(GoWriter writer, BiConsumer<GoMiddlewareGenerator, GoWriter> handlerBodyConsumer) {
+        writeMiddleware(writer, (m, w) -> {
+        }, handlerBodyConsumer);
+    }
+
+    public void writeMiddleware(GoWriter writer, BiConsumer<GoMiddlewareGenerator, GoWriter> fieldConsumer,
+                                BiConsumer<GoMiddlewareGenerator, GoWriter> handlerBodyConsumer) {
+        writer.addUseImports(CONTEXT_TYPE);
+        writer.addUseImports(METADATA_TYPE);
+        writer.addUseImports(inputType);
+        writer.addUseImports(outputType);
+        writer.addUseImports(handlerType);
+
+        Symbol middlwareSym = SymbolUtils.createPointableSymbolBuilder(identifier).build();
+
+        writer.openBlock("type $L struct {", "}", middlwareSym, () -> {
+            fieldConsumer.accept(this, writer);
+        });
+        writer.openBlock("func ($P) ID() string {", "}", middlwareSym, () -> {
+            writer.openBlock("return $S", identifier);
+        });
+        writer.openBlock("func (m $P) $L(ctx $T, in $T, next $T) (\n"
+                        + "\tout $T, metadata $T, err error,\n"
```

Comment:
Minor: it's nice down the road to line up the starting quotes of multi-line components like this.

```
writer.openBlock("foobar\n"
               + "\tbaz"...
```